### DISCUSSION TITLE
insert nested relationship dict

### DISCUSSION THREAD

**Discussion by shaozi** [sqlalchemy/sqlalchemy#7336] [Usage Questions] [answered]
I have a dict that is the data of some nested relationship. Is there an easy way to insert all relations into the tables?

Example:
Suppose the models are defined like the following:

District -o2m-> schools -o2m->students. 

Then I have a dict:

```python
data = { 'district_name': 'a',
    'schools': [{'school_name': 'k1',
        'students': [{'student_name': 's1'}, {'student_name': 's2'}]
    }]
}
```

What I can think of now is to create a District with name only, then create a school, then two Students, then assign them together:

```python
district = District(district_name='a')
school = School(school_name='k1')
s1 = Student('student_name'='s1')
s2 = Student('student_name'='s2')
school.students = [s1, s2]
district.schools = [school]
```

Which is tedious and not generic. 

How can I do something in one shot like this:

```python
district = District(**data)
```

**Comment by CaselIT** [comment, marked as answer]
Hi,

I think we had a similar question some time ago that had a suggestion, but I cannot find it at the moment. @zzzeek do you remember which issue/discussion it was?

In short is used a custom init that using inspection would know how to create the child relationships objects.

Maybe if we find that old answer we could use it to create an example?

**Reply by zzzeek** [reply]
we just had someone that wanted to generate the **models** from a dictionary structure and their solution is at https://github.com/sqlalchemy/sqlalchemy/discussions/7289#discussioncomment-1606411 .

with this one if you want to make it truly generic you would look at inspect(cls).relationships to get the names / related classes of sub-elements.   I'm sure I've done this one too (maybe in the past year) but i think not for a long time.

```py
def __init__(self, **kw):
    mapper = inspect(self).mapper
    for key in mapper.relationships:
        if key in kw:
            kw[key] = mapper.relationships[key].entity.class_(**kw[key])
    super().__init__(**kw)
```

**Reply by CaselIT** [reply]
> we just had someone that wanted to generate the **models** from a dictionary structure and their solution is at [#7289 (comment)](https://github.com/sqlalchemy/sqlalchemy/discussions/7289#discussioncomment-1606411) .

that was not it, something older, I think couple of months back. Well your example above is similar  so never mind

**Reply by shaozi** [reply]
Thank you for pointing to the right direction! It worked!

**Comment by vitordeap** [comment]
Thank you, that's exactly what I needed. 

Updating to newer versions of SQLAlchemy:
```python
def __init__(self, **kw):
    mapper = self.__mapper__
    for key in mapper.relationships.keys():
        if key in kw:
            kw[key] = mapper.relationships[key].entity.class_(**kw[key])

    super().__init__(**kw)

```

**Reply by gsiddhant159** [reply]
Could this be extended to work with dataclasses so that the init params are defined with explicit kw args automatically?

**Reply by zzzeek** [reply]
the purpose of dataclasses is to write an `__init__` method automatically based on annotations.   the recipe here involves using a custom `__init__` that's not like the one dataclasses uses.   I'm not really sure how what you are asking for would be possible
